Enable switching of AWS CLB to NLB without deletion of IngressController object.

miheer opened the pull request:
As an user/admin, I want to switch an OCP Ingress from an AWS Classic Load Balancer to AWS Network Load Balancer. I should be able to do this without having to delete and recreate the IngressController object.
Using this fix one can update the ingress controller object with change of LB type details and get the LB service reflected with the necessary changes without the need for recreation.

Fixes https://issues.redhat.com/browse/NE-865

`pkg/operator/controller/configurable-route/controller.go` We need to mention the key or an error is thrown.
`pkg/operator/controller/ingress/controller.go` Logic added for setting the right endpoint publishing strategy while switching from ELB to NLB or vice-versa.
`pkg/operator/controller/ingress/controller_test.go` Unit tests for switching between NLB and CLB.
`pkg/operator/controller/ingress/deployment_test.go`  We need to mention the key or an error is thrown.
`pkg/operator/controller/ingress/load_balancer_service.go` Logic to check difference between desired and current service for the lb type annotation and proxy protocol annotation and update if the current service with the desired service annotations if necessary.
`pkg/operator/controller/ingress/status.go` Adds progressing status condition logic while switching between NLB and ELB and vice-versa.
` pkg/operator/controller/ingress/status_test.go` Adds a nil as the computeIngressProgressingCondition now also accepts current service as a parameter but for this unit test current lb service is not needed.
`test/e2e/all_test.go` Adds e2e test written for this PR.
`test/e2e/checkinterval_test.go` We need to mention the key or an error is thrown.
`test/e2e/operator_test.go` e2e test for this PR.
`test/e2e/reloadinterval_test.go` We need to mention the key or an error `./reloadinterval_test.go:61:31: k8s.io/apimachinery/pkg/apis/meta/v1.Duration composite literal uses unkeyed fields` is thrown.

miheer commented:
/retest

miheer commented:
/retest

lihongan commented:
/label qe-approved

lihongan commented:
/retest

miheer commented:
/retest

miheer commented:
/retest

gcs278 commented:
/assign @gcs278

gcs278 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -674,6 +674,36 @@ func computeIngressProgressingCondition(conditions []operatorv1.OperatorConditio
+				switch platform.Type {
+				case configv1.AWSPlatformType:
+					if _, ok := service.Annotations[AWSLBTypeAnnotation]; ok {
+						if (ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != "NLB") && (service.Annotations[AWSLBTypeAnnotation] == "nlb") {
```
What about nesting the `if` and doing an `or`? I feel like it might be a little cleaner opposed to having repeated code blocks.
```suggestion
if service.Annotations[AWSLBTypeAnnotation] == AWSNLBAnnotation {
  if ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters == nil || 
     ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS == nil || 
     ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != AWSNetworkLoadBalancer {
 	condition.Reason = "AWSLBTypeChangedToNLBFromClassic"
	condition.Message = message
	condition.Status = operatorv1.ConditionTrue
}
```

gcs278 replied:
In the e2e, we updated ingress controller's `endpointPublishingStrategy` spec, then we wait for the annotation to appear on the service. Here we compare the service annotation with the IC status. Which I think it valid, but what about when the IC.spec is updated, but nor the IC status or the service reflect the change yet? Is that a situation that we should update progressing? Or is that transient enough not to care?

miheer replied:
regarding suggestion for `What about nesting the if and doing an or? I feel like it might be a little cleaner opposed to having repeated code blocks.` it will change the logic. We need to check for the nil value at every level after Provider Params as the go code will crash if nil value is encountered.

miheer replied:
> In the e2e, we updated ingress controller's `endpointPublishingStrategy` spec, then we wait for the annotation to appear on the service. Here we compare the service annotation with the IC status. Which I think it valid, but what about when the IC.spec is updated, but nor the IC status or the service reflect the change yet? Is that a situation that we should update progressing? Or is that transient enough not to care?

I did try to test that but the annotation in the service gets quickly updated and the progressing status quickly becomes false as the conditions to make progressing true become false very quickly.

gcs278 replied:
> regarding suggestion for What about nesting the if and doing an or? I feel like it might be a little cleaner opposed to 
 having repeated code blocks. it will change the logic. We need to check for the nil value at every level after Provider Params as the go code will crash if nil value is encountered.

Agreed, but this is what my suggestion actually does. The code has 3 if conditions with identical outputs. This suggestion is `nil` safe because if the first condition `ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters == nil` is true, then it won't execute the next `ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS == nil`, so on and so forth.

Please double check my work to make sure I'm not missing something, but I'd argue this is more concise and readable. Pseudo Code:
```
if service has NLB annotation {
      if ProviderParameters is nil or ProviderParameters.AWS is nil or ProviderParameters.AWS.type is not NLB {
          set condition to AWSLBTypeChangedToNLBFromClassic
      }
}
```

miheer replied:
@gcs278 cool I got it :) thanks! made the changes.

gcs278 commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -473,17 +473,30 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
-				}
-				if specIdleTimeout != statusIdleTimeout {
+				if specLB.ProviderParameters != nil && specLB.ProviderParameters.AWS != nil && specLB.ProviderParameters.AWS.Type == operatorv1.AWSNetworkLoadBalancer {
+					if statusLB == nil {
```
Optionally, if you do need that logic for `statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters == nil ` you do it this way (but I am unsure because your code had both `!= nil` and `== nil`:
```suggestion
if statusLB.ProviderParameters != nil && statusLB.ProviderParameters.AWS != nil {
  if statusLB.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer || statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters == nil  {
     statusLB.ProviderParameters = specLB.ProviderParameters
     changed = true
}
```

gcs278 replied:
At minimum, consider line 458 and that `statusLB` can never be null, this is a redundant check in each `else if` statement.

gcs278 commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -493,18 +506,30 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
+					if statusLB.ProviderParameters != nil && statusLB.ProviderParameters.AWS != nil && statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters != nil {
+						statusIdleTimeout = statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout
+					}
+					if specIdleTimeout != statusIdleTimeout {
```
Consider initializing statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout to nil and then setting it to specIdleTimeout in this if statement. I think that might be more explicit and readable

```suggestion
if specIdleTimeout != statusIdleTimeout {
  statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout = nil
  if specIdleTimeout.Duration > 0 {
    statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout = specIdleTimeout
  }
  changed = true
}
```

miheer replied:
Actually this part of the code is not related to this PR so did not touch the logic. Nil value was not set in the earlier code as well. I think may be a separate PR can be filed for this to refactor. However I feel setting nil is not needed.

gcs278 replied:
Oh yea now I see the diff. Sorry about that, this can stay as is.

gcs278 reviewed (commented):
Nice PR - mainly just comments on style, otherwise looks good.

miheer commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -473,17 +473,30 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
+					} else if statusLB != nil && statusLB.ProviderParameters != nil && statusLB.ProviderParameters.AWS != nil && statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters == nil {
+						statusLB.ProviderParameters = specLB.ProviderParameters
+						changed = true
+					} else if statusLB != nil && statusLB.ProviderParameters != nil && statusLB.ProviderParameters.AWS != nil && statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters != nil {
```
When we use ingress controller with CLB the ClassicLoadBalancerParameters can be nil or not nil. Eg- The default IC with clb one gets created with  no  ClassicLoadBalancerParameters i.e nil. So having ClassicLoadBalancerParameters is not mandatory for user to set it in the spec of the IC when he wants CLB.

miheer replied:
When we IC has CLB then the ClassicLoadBalancerParameters is not mandatory. Eg- the default IC with CLB has no ClassicLoadBalancerParameters. So it can be set or not set.  So, we need to check for `statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters` for nil and not nil along with the value set by the user in the spec for IC which in this is NLB and then set the provider parameters of the spec to the status which then is used by the lb service code.

gcs278 replied:
Got it thanks for the details, that makes more sense. But checking for something that is nil or not nil is always going to enter one of this `else if` statements, it can't be anything else other than nil or not nil, so what's the point in checking `ClassicLoadBalancerParameters` in the first place? Aren't you effectively doing `exists or not exists` which is always true?

gcs278 replied:
To further my point, if `statusLB.ProviderParameters.AWS != nil`, what logic would make it past this point? `statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters` would always be either `nil` or not `nil`. So the `if else` statements below appear to be dead code, never able to be reached.

miheer commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -961,7 +986,11 @@ func (r *reconciler) ensureIngressController(ci *operatorv1.IngressController, d
 
-	syncStatusErr, updated := r.syncIngressControllerStatus(ci, deployment, deploymentRef, pods.Items, lbService, operandEvents.Items, wildcardRecord, dnsConfig, platformStatus)
+	var currentLBService *corev1.Service
+	if _, currentLBService, err = r.currentLoadBalancerService(ci); err != nil {
```
lbService is the desiredLBService object created w.r.t the IC and not the the current LBService.

gcs278 replied:
It doesn't look like the desiredLBService, it appears it's the currentLBService if I'm reading the code correctly.
`ensureLoadBalancerService` returns the currentLBService as the 2nd return argument, not the `desiredLBSerivce`:
`return true, currentLBService, nil`.

miheer commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -50,7 +50,7 @@ type expectedCondition struct {
 // syncIngressControllerStatus computes the current status of ic and
 // updates status upon any changes since last sync.
-func (r *reconciler) syncIngressControllerStatus(ic *operatorv1.IngressController, deployment *appsv1.Deployment, deploymentRef metav1.OwnerReference, pods []corev1.Pod, service *corev1.Service, operandEvents []corev1.Event, wildcardRecord *iov1.DNSRecord, dnsConfig *configv1.DNS, platformStatus *configv1.PlatformStatus) (error, bool) {
+func (r *reconciler) syncIngressControllerStatus(ic *operatorv1.IngressController, deployment *appsv1.Deployment, deploymentRef metav1.OwnerReference, pods []corev1.Pod, service *corev1.Service, currentLBService *corev1.Service, operandEvents []corev1.Event, wildcardRecord *iov1.DNSRecord, dnsConfig *configv1.DNS, platformStatus *configv1.PlatformStatus) (error, bool) {
```
answered over there

miheer commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -674,6 +674,36 @@ func computeIngressProgressingCondition(conditions []operatorv1.OperatorConditio
 				condition.Reason = "ScopeChanged"
 				condition.Message = message
 				condition.Status = operatorv1.ConditionTrue
+			} else {
```
> Nice PR - mainly just comments on style, otherwise looks good.

Thanks!

miheer commented:
/retest

miheer commented:
/retest

miheer commented:
/retest

miheer commented:
Not sure why the e2e tests are failing.  The details link next to ci tests say ->

~~~~~~~~~

Application is not available
The application is currently not serving requests at this endpoint. It may not have been started or is still starting.

Possible reasons you are seeing this page:

The host doesn't exist. Make sure the hostname was typed correctly and that a route matching this hostname exists.
The host exists, but doesn't have a matching path. Check if the URL path was typed correctly and that the route was created using the desired path.
Route and path matches, but all pods are down. Make sure that the resources exposed by this route (pods, services, deployment configs, etc) have at least one pod running.

~~~~~~~~~

I ran the e2e on local which is working fine.

[miheer@localhost cluster-ingress-operator]$ make test-e2e TEST=TestAWSLBTypeChange
GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestAWSLBTypeChange" ./test/e2e
I0728 16:26:55.283754   81670 request.go:665] Waited for 1.008819894s due to client-side throttling, not priority and fairness, request: GET:https://api.misalunkswitchlb.devcluster.openshift.com:6443/apis/ingress.operator.openshift.io/v1?timeout=32s
=== RUN   TestAWSLBTypeChange
=== PAUSE TestAWSLBTypeChange
=== CONT  TestAWSLBTypeChange
    operator_test.go:1133: load balancer has no "service.beta.kubernetes.io/aws-load-balancer-type" annotation: map[service.beta.kubernetes.io/aws-load-balancer-healthcheck-healthy-threshold:2 service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval:5 service.beta.kubernetes.io/aws-load-balancer-healthcheck-timeout:4 service.beta.kubernetes.io/aws-load-balancer-healthcheck-unhealthy-threshold:2 service.beta.kubernetes.io/aws-load-balancer-proxy-protocol:* traffic-policy.network.alpha.openshift.io/local-with-fallback:]
    operator_test.go:1145: deleted ingresscontroller awslb
--- PASS: TestAWSLBTypeChange (104.52s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/test/e2e	107.041s
[miheer@localhost cluster-ingress-operator]$

miheer commented:
/retest

miheer commented:
@gcs278 nice review :) Thanks a lot!

miheer commented:
@gcs278 i have implemented the changes you asked for. The e2e test also is working fine

[miheer@localhost cluster-ingress-operator]$ make test-e2e TEST=TestAWSLBTypeChange
GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestAWSLBTypeChange" ./test/e2e
I0729 11:11:36.201785   32724 request.go:665] Waited for 1.049038965s due to client-side throttling, not priority and fairness, request: GET:https://api.ci-ln-5wfns0k-76ef8.origin-ci-int-aws.dev.rhcloud.com:6443/apis/controlplane.operator.openshift.io/v1alpha1?timeout=32s
=== RUN   TestAWSLBTypeChange
=== PAUSE TestAWSLBTypeChange
=== CONT  TestAWSLBTypeChange
    operator_test.go:1145: deleted ingresscontroller awslb
--- PASS: TestAWSLBTypeChange (114.65s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/test/e2e	119.247s
[miheer@localhost cluster-ingress-operator]$

gcs278 commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -473,17 +473,10 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
-				}
-				if specIdleTimeout != statusIdleTimeout {
+				if specLB.ProviderParameters != nil && specLB.ProviderParameters.AWS != nil && specLB.ProviderParameters.AWS.Type == operatorv1.AWSNetworkLoadBalancer {
+					statusLB.ProviderParameters = specLB.ProviderParameters
```
Interesting, this looks a lot cleaner. So we didn't need to check if the lbStatus's ProviderParameters to see if it was empty or a `AWSClassicLoadBalancer`?

miheer replied:
The else part will take of this.

miheer commented:
/retest

miheer commented:
/retest

gcs278 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -674,6 +674,29 @@ func computeIngressProgressingCondition(conditions []operatorv1.OperatorConditio
+				case configv1.AWSPlatformType:
+					if _, ok := service.Annotations[AWSLBTypeAnnotation]; ok {
+						if (service.Annotations[AWSLBTypeAnnotation] == AWSNLBAnnotation) &&
+							(ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer) ||
```
Thanks for trying out this change. Do you really need the first stanza that checks for `nil` on `ProvidedParameters` and `AWS` if you rearrange properly?
`(ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer)`?

Couldn't you just do (in this order specifically):
```
if (service.Annotations[AWSLBTypeAnnotation] == AWSNLBAnnotation) &&
(ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters == nil ||
ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS == nil ||
ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer) {
```

The same short circuiting you are using with "don't check `ProviderParameters.AWS.Type` if a prior `&&` is false, is the same with `||` if it is true. It won't attempt to evaluate the next `||` or conditions. Based on the fact you are using `if` conditional short-circuiting, I don't see a reason not to use it with `||`. It is just as `nil` safe as this. Thoughts?

miheer replied:
We need to check if `ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer)`  is not set to NLB(generally for classic LB this is not set but there might be a case where Classic is set when someone wants to specify params in the provider) so for doing that you will need to check if ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS != nil to ensure go does not throw nil error.

I don't think `OR` and  `== nil` will help.

Also we will also need to check the condition if  `(ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters == nil || ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS == nil)`

as for having classic LB  in ingresscontroller it is not mandatory to have the ProviderParameters set in the ingress controller.

gcs278 replied:
Sorry, I don't mean to drag this one out, but I want to make sure we are on the same page. Totally agree with what you said, but just pointing out it could be simplified which makes it a bit more readable.

So do you agree that:
```
if (a.b != nil && a.b.c != nil && a.b.c.d == "x") ||
    (a.b == nil || a.b.c == nil) {
    do something
}
```
the same as this:
```
if a.b == nil || a.b.c == nil || a.b.c.d == "x" {
    do something
}
```
?

miheer replied:
@gcs278  Agree. Thanks. PTAL at the changes I made as per your request.

gcs278 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -50,7 +50,7 @@ type expectedCondition struct {
 // syncIngressControllerStatus computes the current status of ic and
 // updates status upon any changes since last sync.
-func (r *reconciler) syncIngressControllerStatus(ic *operatorv1.IngressController, deployment *appsv1.Deployment, deploymentRef metav1.OwnerReference, pods []corev1.Pod, service *corev1.Service, operandEvents []corev1.Event, wildcardRecord *iov1.DNSRecord, dnsConfig *configv1.DNS, platformStatus *configv1.PlatformStatus) (error, bool) {
+func (r *reconciler) syncIngressControllerStatus(ic *operatorv1.IngressController, deployment *appsv1.Deployment, deploymentRef metav1.OwnerReference, pods []corev1.Pod, currentLBService *corev1.Service, operandEvents []corev1.Event, wildcardRecord *iov1.DNSRecord, dnsConfig *configv1.DNS, platformStatus *configv1.PlatformStatus) (error, bool) {
```
this is a nit pick, but should we really change the existing variable name `service` to `currentLBService`? In context, all of the other variable names are `deployment`, `pod`, `ic`, not `currentDeployment` or `currentIC`. Seems a little inconsistent.

I think the going assumption is that all parameters are "current" objects and not desired. I think it's better to stay consistent with these variable names to avoid confusion.

miheer replied:
@gcs278  done

gcs278 commented:
I just added one more comment, otherwise lgtm. @Miciah said he would review it after I did, so it's ready for you @Miciah now.

Miciah reviewed (commented):
Looks good overall.  I have some suggestions and questions about a few details.

Miciah commented:
/assign

miheer commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -674,6 +674,31 @@ func computeIngressProgressingCondition(conditions []operatorv1.OperatorConditio
+			} else {
+				switch platform.Type {
+				case configv1.AWSPlatformType:
+					if _, ok := service.Annotations[AWSLBTypeAnnotation]; ok {
+						if service.Annotations[AWSLBTypeAnnotation] == AWSNLBAnnotation {
+
+							if ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters == nil ||
+								ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS == nil ||
+								ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type != operatorv1.AWSNetworkLoadBalancer {
+								message := fmt.Sprintf("The IngressController LBType was changed from Classic to NLB")
+								condition.Reason = "AWSLBTypeChangedToNLBFromClassic"
+								condition.Message = message
+								condition.Status = operatorv1.ConditionTrue
+							}
+						}
+					} else if !ok {
+						if ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS != nil && ic.Status.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type == operatorv1.AWSNetworkLoadBalancer {
+							message := fmt.Sprintf("The IngressController LBType was changed from NLB to Classic")
+							condition.Reason = "AWSLBTypeChangedToClassicFromNLB"
+							condition.Message = message
+							condition.Status = operatorv1.ConditionTrue
+						}
+					}
```
Yes it is fast and we don't see it as it is quick.

I did not find any other way.

miheer commented on test/e2e/operator_test.go:
```diff
@@ -1096,6 +1096,89 @@ func TestInternalLoadBalancerGlobalAccessGCP(t *testing.T) {
+	err := wait.PollImmediate(5*time.Second, 5*time.Minute, func() (bool, error) {
+		service := &corev1.Service{}
+		if err := kclient.Get(context.TODO(), controller.LoadBalancerServiceName(ic), service); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
```
just added a check. do you want to delete it ? I think it does not harm to keep a check.

Miciah replied:
It seems to me like this condition should never occur.  Anyway, either the condition cannot occur or it can.  If it cannot occur, then the check is pointless.  If it can but shouldn't occur, then why wouldn't you want the test to log it?

miheer replied:
OK will remove it.

miheer commented:
/retest

miheer commented:
@Miciah PTAL

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -498,18 +491,30 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
 					}
-					statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout = v
-					changed = true
 				}
```
Did you consider this suggestion?  I think it makes the logic much simpler.

Miciah commented:
Just so it doesn't get lost due to the way GitHub threads comments, I'll note that the following questions are outstanding:

* https://github.com/openshift/cluster-ingress-operator/pull/790#discussion_r966467715
* https://github.com/openshift/cluster-ingress-operator/pull/790#discussion_r966469721

miheer commented:
@Miciah PTAL. Thanks a lot for the suggestions!

miheer commented:
/retest

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -498,19 +540,31 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
+						statusLB.ProviderParameters.AWS.ClassicLoadBalancerParameters.ConnectionIdleTimeout = v
+						changed = true
+					}
+				}*/
```
No need to keep the commented-out code.

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -477,39 +477,46 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
 				lbType = specLB.ProviderParameters.Type
 			}
 			switch lbType {
+			//case operatorv1.AWSLoadBalancerProvider:
```
No need to keep this commented-out code.

miheer commented:
@Miciah PTAL

miheer commented:
/retest

Miciah commented:
/lgtm

miheer commented:
@gcs278 can you please /approved ?

gcs278 commented:
/lgtm

Miciah commented:
/approve

jldohmann commented:
/label docs-approved

candita commented:
/retest-required

lihongan commented:
/label qe-approved

CFields651 commented:
/label px-approved

miheer commented:
/retest-required

candita commented:
Failing invariants reported in https://coreos.slack.com/archives/C01CQA76KMX/p1664216773179949, as this started after a change in openshift/origin made for https://issues.redhat.com//browse/TRT-474

[sig-network-edge] ns/openshift-authentication route/oauth-openshift disruption/ingress-to-oauth-server connection/new should be available throughout the test
[sig-network-edge] ns/openshift-console route/console disruption/ingress-to-console connection/new should be available throughout the test

Miciah commented:
/override ci/prow/e2e-gcp-ovn-serial
